### PR TITLE
Poro-34B-chat tokenizer support

### PR DESCRIPTION
**Pull request by ezosa**
Implemented pre-tokenizer support for [Poro-34B-chat](https://huggingface.co/LumiOpen/Poro-34B-chat).

- Added tokenizer type for Poro-34B-chat in `convert-hf-to-gguf-update.py`
- Added the chkhsh for Poro-34B-chat in `convert-hf-to-gguf.py`
- Added LLAMA_VOCAB_PRE_TYPE_PORO enum to `llama.h`
- Added pre-tokenizer regex for LLAMA_VOCAB_PRE_TYPE_PORO to `llama.cpp`
- Ran ./tests/test-tokenizer-0 ./models/ggml-vocab-Poro-34B-chat.gguf. Tests passed.

Related to PR #7328 since Poro and Viking share the same pre-tokenizer regex

### CODE REVIEW

**Comment by akx** [comment]
@ezosa Are you seeing the same issues as in https://github.com/ggerganov/llama.cpp/pull/7328#discussion_r1620786981? 🤔

**Comment by ezosa** [comment]
> @ezosa Are you seeing the same issues as in [#7328 (comment)](https://github.com/ggerganov/llama.cpp/pull/7328#discussion_r1620786981)? 🤔

Weirdly, no. My tests all passed. Maybe the failed test is specific to Viking? I'll have a look at Viking soon.

```
src: 'Hello, y'all! How are you 😁 ?我想在apple工作1314151天～'
res: 'Hello, y'all! How are you 😁 ?我想在apple工作1314151天～'
tok: 17720 35 356 90701 24 2888 564 569 11892 234 2076 13217 37414 7359 21264 55110 1688 1581 45843 29066 65074 263 

src: 'ied 4 ½ months'
res: 'ied 4 ½ months'
tok: 907 802 51074 5481 

src: 'w048 7tuijk dsdfhu'
res: 'w048 7tuijk dsdfhu'
tok: 72235 2928 1158 507 72043 32710 3128 3836 

src: 'нещо на Български'
res: 'нещо на Български'
tok: 40411 12118 921 7866 24106 24892 1953 6197 13534 19610 

src: 'កាន់តែពិសេសអាចខលចេញ'
res: 'កាន់តែពិសេសអាចខលចេញ'
tok: 19523 233 104963 252 52087 244 19523 248 52087 235 19523 255 19523 139 19523 264 52087 234 19523 264 19523 119 104963 238 19523 234 19523 260 19523 238 52087 234 19523 242 

src: '🚀 (normal) 😶‍🌫️ (multiple emojis concatenated) ✅ (only emoji that has its own token)'
res: '🚀 (normal) 😶‍🌫️ (multiple emojis concatenated) ✅ (only emoji that has its own token)'
tok: 4318 259 233 365 16007 32 11892 138 102753 117264 128 26036 365 66533 2953 106742 65851 708 32 38132 238 365 5864 88269 451 773 920 1974 7023 32 

Tests passed
```

**Comment by jonabur** [comment]
The regex @ezosa used is slightly different than the one I used, so that likely explains the difference in test performance.

**Comment by jonabur** [comment]
Any hope on getting this merged?

File: convert-hf-to-gguf-update.py [hunk 1]
```diff
@@ -83,6 +83,7 @@ class TOKENIZER_TYPE(IntEnum):
     {"name": "jina-v2-es",     "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/jinaai/jina-embeddings-v2-base-es", },
     {"name": "jina-v2-de",     "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/jinaai/jina-embeddings-v2-base-de", },
     {"name": "smaug-bpe",      "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/abacusai/Smaug-Llama-3-70B-Instruct", },
+    {"name": "Poro-34B-chat",  "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/LumiOpen/Poro-34B-chat", },
```
Comment (ggerganov) [inline convert-hf-to-gguf-update.py:86]: Let's change the name to `poro-chat`:

```suggestion
    {"name": "poro-chat",  "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/LumiOpen/Poro-34B-chat", },
```

**Review by ggerganov — APPROVED** [review]

**Review by ezosa — COMMENTED** [review]
Changed Poro-34B-chat to poro-chat in the relevant files

File: convert-hf-to-gguf-update.py [hunk 2]
```diff
@@ -83,6 +83,7 @@ class TOKENIZER_TYPE(IntEnum):
     {"name": "jina-v2-es",     "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/jinaai/jina-embeddings-v2-base-es", },
     {"name": "jina-v2-de",     "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/jinaai/jina-embeddings-v2-base-de", },
     {"name": "smaug-bpe",      "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/abacusai/Smaug-Llama-3-70B-Instruct", },
+    {"name": "poro-chat",  "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/LumiOpen/Poro-34B-chat", },
```
Comment (ggerganov) [inline convert-hf-to-gguf-update.py:86]:
```suggestion
    {"name": "smaug-bpe",      "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/abacusai/Smaug-Llama-3-70B-Instruct", },
    {"name": "poro-chat",      "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/LumiOpen/Poro-34B-chat", },
```

File: llama.cpp [hunk 3]
```diff
@@ -4705,7 +4705,11 @@ static void llm_load_vocab(
             } else if (
                 tokenizer_pre == "smaug-bpe") {
                 vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_SMAUG;
-            } else {
+            } else if (
+                tokenizer_pre == "poro-chat") {
+                vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_PORO;
+            } 
+            else {
```
Comment (ggerganov) [inline llama.cpp:4712]:
```suggestion
            } else {
```

**Pull request merged**